File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        curr = ListNode()
        dummy = curr
        flag = 1
        while head:
            revNode, revNodenext, nodesLessthanK = self.reverseSubList(head, k)
            head = revNodenext
            curr.next = revNode
            
            n = k + 1
            while n > 1 and nodesLessthanK:
                curr = curr.next
                n -= 1
            
        return dummy.next


    def reverseSubList(self, head: Optional[ListNode], k: int) -> [Optional[ListNode],  Optional[ListNode], int]:
        if k == 1:
            return [head, head.next, 1]
        
        if head is None or head.next is None:
            return [head, head.next, 0]
    
        node, nodeNext, nodesLessthanK = self.reverseSubList(head.next, k - 1)
        if nodesLessthanK:
            head.next.next = head
            head.next = None
            return [node, nodeNext, nodesLessthanK]
        return [head, nodeNext, nodesLessthanK]
        

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode dummy = new ListNode();
        ListNode current = head;
        
        ListNode prev_subList = dummy;
        ListNode prev = dummy;
        int length = 0;

        while (current != null) {
            length++;
            current = current.next;
        }

        current = head;
        dummy.next = current;
        
        for (int i = 0; i < length / k; i++) {
            int counter = k;
            prev = prev_subList;
            
            while (counter > 0) {
                ListNode next = current.next;
                current.next = prev;
                prev = current;
                current = next;
                counter--;
            }

            prev_subList.next.next = current;
            ListNode nextPrev = prev_subList.next;
            prev_subList.next = prev;
            prev_subList = nextPrev;
        }
        return dummy.next;
    }
}